### DISCUSSION TITLE
On an H3, what do feed_in_r/s/t and inv_power_r+s+t exactly indicate?

### DISCUSSION THREAD

**Discussion by xrad** [nathanmarlor/foxess_modbus#583] [Q&A]
What the title says. Obviously the first impulse is to say it's the feed in to the grid, but it's not. For instance right now my battery is still charging with 600 W. I can see the usual ~40 W coming in from the grid, so there is definitely no feeding to the grid happening. Still, feed_in_r+s+t amounts to about 70 W. What is going on?

While there, could anybody tell me what exactly inv_power_r+s+t encompasses? 

Sorry for this possibly n00by types of questions, but I really was not able to find anything reliable on that, just a ton of forum posts with varying degrees of accuracy. I'm also not sure to what extent the inverter model and the master firmware plays into this.

**Comment by canton7** [comment]
@TonyM1958 put together a good picture: https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/How-does-it-work%3F

**Reply by canton7** [reply]
(For legacy reasons, `inv_power` on the H3 is called `rpower` on that image)

**Reply by xrad** [reply]
Thanks for the super-speedy response. This looks like a treasure trove indeed!

**Reply by canton7** [reply]
So yes, feed_in is the power going out through the meter which is connected to your H3 (normally Chint or similar). This is the value which the H3 uses to balance its output, i.e. most of the time the inverter is working to keep the power through the Chint at 0 (or a slight import typically), by providing output power equal to your house load.

**Reply by xrad** [reply]
Unfortunately the image or the text do not really explain the relation of the Feed-in Power and GridConsumptionPower to the CT meter. I know that the names seem to be very self-explanatory, but as I said above: in situations where there is clearly nothing going out to the grid (as per CT measurment, which should be most accurate) - how can Fedd-in Power than be anything but 0 W?  

Edit: there was an overlap, I did not read you last post when I wrote this.

**Reply by canton7** [reply]
Note that H3s don't use a CT clamp (that wouldn't work for 3-phase power). They need to use a smarter meter, such as a Chint.

Feed-in power is the power going through the Chint. 

How are you determining that "as per CT measurement"?

**Reply by xrad** [reply]
Sorry for possibly fuzzy language, all of this is quite new to me. Yes, a DTSU666 is connected. So here is a real-life snapshot:

pv_solar = 1600 W
battery_charge_r/s/t = 1300 W
grid_ct = - 22 W
grid_consumption_power = 80 W
feed_in_r/s/t = 50 W

Why is there a positive feed-in? What does it indicate?

**Reply by canton7** [reply]
So. 

Firstly, grid_ct_r/s/t is the value read from your Chint meter. feed_in_r and grid_consumption_r are directly derived from grid_ct_r:

| grid_ct_r | feed_in_r | grid_consumption_r |
| ------------- | ------------- | --- |
| 0 | 0 | 0 |
| > 0 | grid_ct_r | 0 |
| < 0 | 0 | -grid_ct_r |

As you can see, feed_in_r and grid_consumption_r are just different ways of looking at grid_ct_r: if grid_ct_r is positive when you're feeding in, and if it's negative you're importing.

So, you should never have a case where feed_in_r and grid_consumption_r are both non-zero. If feed_in_r / grid_consumption_r are non-zero, then they should be the same as grid_ct_r (negated in the case of grid_consumption_r).

The same applies to the other two phases, s and t.

However, this is all per-phase. It's perfectly possible for your inverter to be importing on one phase and exporting on another. This means that directly summing the phases for feed_in and grid_consumption doesn't necessarily make sense. For example:

| | R | S | T | Sum |
| --- | --- | --- | --- | --- |
Grid CT | 10W | 10W | -10W | 10W |
Feed-in | 10W | 10W | 0W | 20W |
Grid Consumption | 0 | 0 | 10W | 10W |

See https://github.com/nathanmarlor/foxess_modbus/discussions/456 for a longer discussion.

**Reply by xrad** [reply]
I will need to ponder this for a while. Thanks a ton for taking time to explain all of this.

**Comment by TonyM1958** [comment]
The main thingyou need to understand here is net metering - this does not apply in all regions, but mostly it does and it is how H3 calculates it's power, load etc.

Your CHINT does not measure Grid Consumption and Feed In separately - these can be created within the integration but it's not how the inverter works.

Meter Power is the sum of Meter Power R + Meter Power S + Meter Power T.

As long as this sums to zero, there is no net import or export. Once you do this sum, Feedin is the value when +ve and Grid Consumption is the value when -ve. There is never a time when both of these have a non-zero value and there is no such thing as Grid Consumption RST or Feedin RST.

So, in the example table above, the sum of Grid CT / Meter Power is 10W, Grid Consumption is 0W and Feedin is 10W.

**Reply by xrad** [reply]
The pieces start falling into place, thanks. I was aware (kind of) of what you explained, but always assumed the integration entities where an 1:1 representation of modbus registers. So I went to the integration's source code and saw that grid consumption and feed in are indeed _calculated_ from the raw meter registers as you described. I got my initial knowledge from a number of possibly older forum posts - perhaps at the time they were written the integration simply wasn't as advanced as it is today.

One final question remains: I would be interested in calculating the inverter loss (should be ~ 200 W for an H3.) Currently I'm doing that by summing up all possible input power values (solar, grid, battery discharge) and subtracting all possible output power values (inverter output, battery charge.) This more or less ends up in the right ballpark, but I think it is a little too much. Any idea on that one?

**Reply by TonyM1958** [reply]
This is how I calculate system losses:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/5035a862-a279-4f1a-869e-3a3b509c9620)

I put the losses, input and output power through Riemann sum and utility meters to track daily energy loss:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/16a11d6d-d1d6-4048-96eb-fbb0ccee3e8b)

I generally get over 90% on H1 but I have seen efficiencies as low as 75% on H3. Fox was supposed to be developing a firmware update that improved the efficiency of H3 as people have complained about it.

**Reply by xrad** [reply]
That's interesting, thanks for sharing. When would sensor.rst/power ever become negative? I just checked the past few days on my system and these values were never negative. (Just to be clear: this is inv_power_rst, right?)

**Reply by TonyM1958** [reply]
The inverter draws power from the grid to charge the batteries.

**Reply by xrad** [reply]
Aha, so your sensor.(rst)power is the meter, not the inverter power. Then it makes more sense.

**Reply by xrad** [reply]
So technically you could replace your rpower/spower/tpower calculations by feedin and consumption provided by the integration, which already does the same calculations, correct?

**Reply by TonyM1958** [reply]
No and no. RST Power are the 3 phase connections to the inverter. They are not the same as Meter Power.

**Reply by xrad** [reply]
Well ok, sorry for the noise. I had looked earlier for this but apparrently not closely enough. Now I do see negative power values, albeit very small ones.
